### ISSUE TITLE
Feature Request: Add option to disable attribute sorting

### ISSUE THREAD

**Issue by panoply** [prettydiff/prettydiff#614]
Currently, alphabetical sorting is enforced, from what I can tell. Would be great if alphabetically sorting markup tag attributes was optional or we could define the hierarchy.

**Comment by prettydiff** [comment]
I have attribute sorting on by default because the HTML and XML specifications state that all attributes of a given tag are equal peers without precedence, and so there is no harm from sorting.  The benefit is some amount of consistency when reading the code (my opinion).  **I could write an option to turn this off though.**

I do have attribute sorting disabled in cases where an attribute opens up to a nested language, like JSX attributes escaping to JavaScript.  I also disable sorting when a tag contains attributes that are part of some template scheme, because in this case order may very well matter.

---

Sorting is something I do in the parser.  It seemed like a good idea, at the time, that the parse tree should exactly represent the interpreted code while Pretty Diff was limited to formatting the result of the parse tree.  It would have been so much easier to mutate the parse tree at beautification time, because the formatted code output lacks so much of the sophistication that goes into all the details of parsing.

I am more than happy to write up additional sort mechanisms, but it is a lot of work.  Since it is a lot of work I need you to be extremely precise in what you would like with with hierarchy definitions in code examples of input and different mutated output.

**Comment by panoply** [comment]
Developers from what I have observed will generally order attributes according to their level of importance. [This specification](https://codeguide.co/#html-attribute-order) by [mdo](https://github.com/mdo) is a good reference point here to order in which most developers would write HTML attributes. 

The idea of formatting the attributes according to a predefined hierarchy would allow developers to sort in accordance to what they see as a level of importance. Passing in an array option could suffice nicely here, where the values would set the starting order, eg:

```javascript
{
  mode:  'beautify',
  lexer: 'markup',
  sort_attributes: [ 
    'id', 
    'type',
    'src'
    'class', 
    'data-*', 
    'aria-*' ,
    'data-*-end',
    'required'
  ]
}
```
The `sort_attributes` option could default to `alphabetical` on attributes not defined in the array and when the `sort_attributes` first value is of type boolean and set to `false` sorting could be skipped altogether.

Alternatively and given that no prettydiff beautify exposed option (from my knowledge) accepts an array, it could instead accept a `string` with comma separated values. Following common rules, the values `"never"` would skip sorting whereas `"always"` would default to alphabetical sorting, eg: 

```javascript
{
  mode:  'beautify',
  lexer: 'markup',
  sort_attributes:  'id, type, src, class, data-*, aria-*, data-*-end, required'
}
```

The attributes using an asterisk in the example would denote a wildcard. Using the above example values `aria-*`, `data-*` and `data-*-end` would result in attributes formatting as:

```html
<button 
    type="button" 
    id="foo" 
    class="bar" 
    data-foo-id="hello"
    data-foo-name="world"
    aria-min="0" 
    aria-max="100" 
    data-novalue-end>
```

### Helpful Reference

A real world example of this implementation is the **posthtml-attrs-sorter** which could probaly be improved upon but is good starting point, see [here](https://github.com/mrmlnc/posthtml-attrs-sorter/blob/master/index.js).

**Comment by panoply** [comment]
> I am more than happy to write up additional sort mechanisms, but it is a lot of work. 

The last thing you need is to make a refactor and open up other areas to issues, so if the juice is not worth the squeeze on this one it can be skipped.

**Comment by prettydiff** [comment]
I am publishing the fix to this right now.  To solve for this I added two new options:

* attribute_sort
* attribute_sort_list

The first is a boolean with a default of false, so attribute sorting is no longer a thing for any language by default.  Even when set to true I continue to disable it where I think it will be dangerous.  The second attribute is a string of comma separated values.  This second option will not work if the former is value false.

I didn't really add any new logic to the Pretty Diff repo to expose these features.  The work was all done in the parser.  I just exposed the options, and the build makes the options visible everywhere, and there is a small piece of code that automagically wires up options to the parser's API, so really everything is almost fully automated.  The biggest effort was just updating tests.
